luavm/src/lib/lua_values/lua_closure.rs

```diff
@@ -1,88 +1,139 @@
 use crate::{stdlib::StdFunction, Vm};
-use gc::GcCell;
+use gc::{Finalize, Gc, GcCell, Trace};
 use luacompiler::bytecode::Function;
 
+impl Finalize for Box<LuaClosure> {}
+unsafe impl Trace for Box<LuaClosure> {
+    custom_trace!(_this, {});
+}
+
 /// Represents a closure in Lua.
 #[derive(Trace, Finalize)]
-pub struct LuaClosure {
+pub struct UserFunction {
     index: usize,
     reg_count: usize,
     param_count: usize,
     args_count: GcCell<usize>,
     args_start: GcCell<usize>,
-    #[unsafe_ignore_trace]
-    // if the handler is set to some function defined in the vm, then when the closure is
-    // processed as part of a call instruction, the handler is called instead of `vm.eval`
-    handler: Option<fn(&mut Vm)>,
 }
 
-impl From<&Function> for LuaClosure {
-    fn from(func: &Function) -> LuaClosure {
-        LuaClosure {
-            index: func.index(),
-            reg_count: func.reg_count(),
-            param_count: func.param_count(),
+impl UserFunction {
+    pub fn new(index: usize, reg_count: usize, param_count: usize) -> UserFunction {
+        UserFunction {
+            index,
+            reg_count,
+            param_count,
             args_count: GcCell::new(0),
             args_start: GcCell::new(0),
-            handler: None,
         }
     }
 }
 
-impl From<&StdFunction> for LuaClosure {
-    fn from(func: &StdFunction) -> LuaClosure {
-        LuaClosure {
-            index: 0,
-            reg_count: func.reg_count(),
-            param_count: func.param_count(),
-            args_count: GcCell::new(0),
-            args_start: GcCell::new(0),
-            handler: Some(func.handler()),
-        }
+impl LuaClosure for UserFunction {
+    fn index(&self) -> usize {
+        self.index
     }
-}
 
-impl LuaClosure {
-    pub fn new() -> LuaClosure {
-        LuaClosure {
-            index: 0,
-            reg_count: 1,
-            param_count: 0,
-            args_count: GcCell::new(0),
-            args_start: GcCell::new(0),
-            handler: None,
-        }
+    fn args_count(&self) -> usize {
+        self.args_count.borrow().clone()
     }
 
-    pub fn index(&self) -> usize {
-        self.index
+    fn set_args_count(&self, count: usize) {
+        *self.args_count.borrow_mut() = count;
+    }
+
+    fn args_start(&self) -> usize {
+        self.args_start.borrow().clone()
+    }
+
+    fn set_args_start(&self, count: usize) {
+        *self.args_start.borrow_mut() = count;
+    }
+
+    fn reg_count(&self) -> usize {
+        self.reg_count
     }
 
-    pub fn args_count(&self) -> usize {
+    fn param_count(&self) -> usize {
+        self.param_count
+    }
+
+    fn call(&self, vm: &mut Vm) {
+        vm.eval();
+    }
+}
+
+#[derive(Trace, Finalize)]
+pub struct BuiltinFunction {
+    #[unsafe_ignore_trace]
+    handler: fn(&mut Vm),
+    reg_count: usize,
```

Comment:
Does a builtin function need registers?

Reply:
Some functions might use the `_ENV` register, so `reg_count` would always be 1, I guess I can remove the field, and just return 1 from `reg_count()`.

Reply:
I pushed a commit for this.

Reply:
Does the `Vm` object not have a reference to `_ENV`? Then you wouldn't have to worry about registers at all.